cosima_cookbook/database.py

```diff
@@ -759,12 +761,40 @@ def index_experiment(files, session, expt, client=None):
 
     update_metadata(expt, session)
 
-    results = []
-
-    results = [index_file(f, experiment=expt, session=session) for f in tqdm(files)]
+    # Short-cut if no files are specified
+    if len(files) == 0:
+        return 0
+
+    if nfiles is None:
+        # Default to a "chunk size" of 1000 or number of files, whichever
+        # is smaller
+        nfiles = min(1000, len(files))
+
+    def chunks(setlist, n):
+        """Yield successive n-sized chunks from setlist.  Last yielded chunk may have fewer
+        than n elements: len(setlist) does not have to be an integer multiple of n"""
+        lst = list(setlist)
+        for i in range(0, len(lst), n):
+            yield lst[i : i + n]
+
+    nindexed = 0
+
+    # Cap the maximum number of files to index before committing to keep memory use
+    # under control and make indexing less affected by errors
+    for fileschunk in chunks(files, nfiles):
+        results = [
+            index_file(f, experiment=expt, session=session) for f in tqdm(fileschunk)
+        ]
+        try:
+            session.add_all(results)
+        except:
+            raise
```

Comment:
Not strictly necessary, you can just have `try...finally` for this same behaviour. However, is it worth having a failing chunk not compromise the rest of the files in the experiment, i.e. handle some exceptions here?

Also, should `session.commit()` be in the `finally` block? I guess I can't remember the `index_experiment`/`index_file` interaction anyway, and who's responsible for what!

Reply:
I wasn't sure about the best approach TBH. Previously we've not had any `try`/`except` machinery around this. 

With being done in chunks it maybe make sense to just log/print the error and continue with the next chunk which is I guess what you're suggesting, but for selected exceptions.

Reply:
That's true, there are just a few exceptions logged in `index_file`, so maybe there's nothing that even bubbles up to here?

Reply:
I've not looked. Logging an error and continuing makes it more robust without affecting the DB. The downside is potentially missing some files in the DB. At first glance this doesn't seem so bad, but now I think about it, I believe queries will quite happily return an `xarray.DataArray` with non-monotonic time axes and users would not know if they didn't check. That seems bad, so perhaps the priority is to ensure errors are dealt with before more indexing is done? Another option is to set a flag in the experiment or variable to indicate there are some issues with missing data?

In any case it is starting to sound out of scope, and I'd probably just be happy to remove the `try/except, revert to the previous behaviour and cover the rest of it in an [SEP Field](https://en.wikipedia.org/wiki/Somebody_else%27s_problem).

Reply:
Agreed that it's probably out of scope to handle that in this PR. I think it would be better to have the more robust/faster indexing for the moment, and maybe if we see significant trouble in the logs it's worth thinking about how to deal with those files/experiments.